### ISSUE TITLE
How to add zend_loader extension not included in the list?

### ISSUE THREAD

**Issue by wangping710** [cloudfoundry/php-buildpack#51]
Hi,
how could I add it?

Thanks!

**Comment by dmikusa-pivotal** [comment]
Hello!  Can you provide some more details about this?  I see [zend_loader](http://framework.zend.com/manual/1.12/en/zend.loader.html), but it just seems like just PHP code (if I'm missing something here, please let me know).  

The *only* thing that we package with the build pack are C extensions to PHP, because it would be really difficult for a developer to include those.  If you need PHP code, you should package it with your app or use Composer to pull it in.

**Comment by hairmare** [comment]
This could possibly reference the [Zend Guard Loader](http://files.zend.com/help/Zend-Guard/content/installing_zend_guard_loader.htm) (called zend_loader internally on Zend Server for instance).
The code @dmikusa-pivotal referenced is just a small part of Zend Framework that incidentally has the same name.
@wangping710 Can you confirm either theory?

**Comment by dmikusa-pivotal** [comment]
If it is Zend Guard, I don't think we could include that.  It appears to be a commercial product and we'd need Legal to OK it first.

While it's not likely to be pulled into the build pack, if you need this you could include it with a [custom extension](https://github.com/cloudfoundry/php-buildpack/blob/master/docs/development.md#extensions).  The extension would just need to download the Linux archive from Zend, extract the binary extension and copy it into the extension directory.  If you have any questions about this, let me know.

**Comment by hairmare** [comment]
Just FYI. zend_loader should fly through legal with full sails. It's license was written permitting it being packaged in just this way.

It is part of Zends solution to encrypting closed-source php code that comes complete with enterprisey licensing server parts and whatnot. Zends clients use it to ship proprietary php code and do things like per machine or seat licensing. The loader is the part the decrypts encrypted bytecode that was deployed to any server that supports it. If I recall correctly, It runs that on the targets zend engine bypassing phps internal bytcode compiler.

Basically by including support you would expand cfs market to a handful of Zend customers that want to deploy encrypted php code. For such customers this might make cf an attractive alternative to deploying Zend Server instances.

**Comment by dmikusa-pivotal** [comment]
Interesting.  I'll admit once I saw it was a commercial product and didn't see a readily available link to their licensing terms, I stopped investigating and assumed it would be possible to include.  While I still can't find a link to their licensing terms for the product, I did see a FAQ entry that seemed to indicate what you said, so perhaps we could include this.

I guess we'll have to wait and see if this is what's being requested and if so, how we'd go about getting legal approval.  I'm less optimistic that it'll be quick, but regardless we'll pursue if given user interest.

@hairmare is this something you're using or would like to use?

**Comment by dmikusa-pivotal** [comment]
Oh, also this seems to only be available for PHP 5.4.  Is that correct or am I missing something?  I didn't see any links for PHP 5.5 or 5.6 on zend's site.

**Comment by hairmare** [comment]
I don't have a use case for zend guard at the moment. I was just clarifying its context since you seem to be more of a python man and php has some pretty funky internals when it gets down to this level.

I may well be that it is only available for php 5.4 (and previously for 5.3). There where some large changes at low levels in 5.5 and people switched to other solutions in the bytecode cache area. I'm not sure who is currently market leader in providing php encryption solutions for 5.5 and up though (or if they even still exist). There used to be other players in this market before Zend Guard and before 5.3.

Personally I'm glad I got away from having to encrypt the code I ship.

Maybe bluemix already does this (it seems they have zend as partner somewhere), Zend don't seem to have it in their build pack at [zendtech/zend-server-php-buildpack](https://github.com/zendtech/zend-server-php-buildpack) but I didn't and won't test those facts. Their buildpack might be the more appropriate for addressing this problem anyhow.

**Comment by wangping710** [comment]
sorry， i not reply everybody in time 

i need Zend Guard Loader，studying to how put it in the buildpack with custom extension .

is there a more simple way，please？

**Comment by flavorjones** [comment]
Reopening this; will close duplicate #52.

**Comment by flavorjones** [comment]
@wangping710 - Can you please confirm our understanding of your requirements? @dmikusa-pivotal and @hairmare asked a few questions above. Thank you for using Cloud Foundry!

**Comment by dmikusa-pivotal** [comment]
@wangping710 - including Zend Guard Loader with the build pack would be tricky.  It's a commercial product, so we'd have to get approval from the legal team to do so.  Rather than have you wait on this, which may take quite some time, I'd suggest going with a custom extension to the build pack.  The extension should be fairly simple,  just needing to download the file from Zend, extract it and put the files into the correct place.

You're right that it shouldn't be this difficult to add in a custom third party PHP extension.  Unfortunately, I haven't had time to work on #33 which is a request to be able to easily add custom PHP extensions.

**Comment by wangping710** [comment]
hi，

We use zendloader is to let CF run PHP application encrypted, so we need is the Zend Guard Loader@dmikusa-pivotal 、 @hairmare and @flavorjones

Custom extension provides about what use? @dmikusa-pivotal 

thanks！

**Comment by dmikusa-pivotal** [comment]
@wangping710 - If you look at the NewRelic build pack extension, you can see an example of what I'm talking about here.  That extension reaches out to the internet to download and install the NewRelic software.  Since NewRelic is a commercial product and we are not able to redistribute it, we have to install it on-demand.  You could do the same thing here with Zend Guard.

Since Zend Guard is something that might be generally helpful to build pack users, I think it would be worth adding to the build pack even if it's just an extension that downloads and installs the software (i.e. we don't redistribute it with the bp).  That said, it might take some time to get around to adding this.  If it's something you need soon, you'd be better off writing the extension your self.

@flavorjones - what do you think about adding support for this as an extension that downloads and installs the software from the Internet?  On the plus side, we wouldn't be distributing the product so I don't think there would be any legal issues.  On the negative side, support for offline users would be much harder / not possible.  If you think that's OK, I'll look at writing up an extension I think it should be pretty simple.

**Comment by hairmare** [comment]
@dmikusa-pivotal Could you allow users to push the downloaded and unpacked extension in ``.bp-config/`` somewhere? I might be missing something, but it seems that doing so could take care of the offline case.

**Comment by dmikusa-pivotal** [comment]
@hairmare - That's certainly a possibility.  There's actually a similar ticket for a general solution to adding custom extensions.  See #33.

**Comment by wangping710** [comment]
@dmikusa-pivotal - I trying to add Zend Guard Loader with custom extension way,but still hope PHP buildpack itself as soon as possible to support this function.

thanks!

**Comment by flavorjones** [comment]
We won't add this extension right now; we hope to make buildpacks more extensible soon, which would allow you to add custom extensions more easily.

I will leave this ticket open for now. Please let us know if you need help building your own extension.
